src/analysis/functions/infer_callconv.rs

```diff
@@ -0,0 +1,280 @@
+//! Infers the calling convention of each function.
+
+use frontend::imports::ImportInfo;
+use frontend::radeco_containers::RadecoModule;
+use middle::dce;
+use middle::ir;
+use middle::regfile::SubRegisterFile;
+use middle::ssa::cfg_traits::*;
+use middle::ssa::ssa_traits::*;
+use middle::ssa::ssastorage::SSAStorage;
+
+use petgraph::prelude::*;
+use petgraph::visit::{DfsPostOrder, VisitMap, Walker};
+
+use std::collections::{hash_map, HashMap, HashSet};
+use std::iter::{Extend, FromIterator};
+
+/// For every function, record which registers it reads and which registers it
+/// preserves.
+///
+/// This analysis is super conservative; for example, if a function preserves a
+/// register by pushing it onto the stack and popping it back right before
+/// returning, it is considered to be read and not preserved because we can't
+/// guarantee that that stack location is never subsequently read or modified.
+pub fn go(rmod: &mut RadecoModule) -> () {
+    let mut a = Analyzer::new();
+    a.go(rmod);
+    for (addr, cc) in &a.call_convs {
+        let rfn = &rmod.functions[addr];
+        eprintln!("{:?} (@ {:#X}):", rfn.name, rfn.offset);
+        eprintln!("  reads: {:?}", cc.reads);
+        eprintln!("  preserves: {:?}", cc.preserves);
+    }
+}
+
+// TODO: index in SubRegisterFile?
+type Reg = String;
+
+#[derive(Debug, Clone)]
+struct CallingConvention {
+    /// (potential) parameters
+    reads: HashSet<Reg>,
+    /// Callee-saved registers
+    preserves: HashSet<Reg>,
+}
+
+impl CallingConvention {
+    fn is_compatible_with(&self, other: &CallingConvention) -> bool {
+        self.reads.is_subset(&other.reads) && self.preserves.is_superset(&other.preserves)
+    }
+}
+
+struct Analyzer {
+    /// Map from function addresses to their calling convention
+    call_convs: HashMap<u64, CallingConvention>,
+}
+
+impl Analyzer {
+    fn new() -> Analyzer {
+        Analyzer {
+            call_convs: HashMap::new(),
+        }
+    }
+
+    /// Calls `patch_fn`, `dce::collect`, and `analyze_fn` on every function,
+    /// callees first
+    fn go(&mut self, rmod: &mut RadecoModule) -> () {
```

Comment:
`analyze()` better than `go` maybe?

Reply:
It does more than just analysis though; it also patches call sites.